### ISSUE TITLE
Bulk replacement of uninteresting values with zero

### ISSUE THREAD

**Issue by DRMacIver** [HypothesisWorks/hypothesis#1098]
Note: This is part of #1093, where I encourage people to work on the shrinker. As such I will not be working on it myself and encourage you to take a look at it if it takes your fancy.

I was looking at `test_shrink_quality.py` and in particular `test_minimize_multiple_elements_in_silly_large_int_range` and noticed something interesting: It's very laboriously going through each of its integers one by one and replacing them with zero.

It would be nice if it could do this in one fell swoop by trying a large scale reduction that replaces every one of them with zero.

The approach I'd probably start with is:

* Take a look at the `zero_example` shrink pass ([somewhere around here](https://github.com/HypothesisWorks/hypothesis/blob/master/hypothesis-python/src/hypothesis/internal/conjecture/engine.py#L2256))
* When it successfully zeroes a draw, replacing it with its minimal example, rather than just continuing on with the loop, try replacing some large number of similar examples with that drawn one.
* A good first pass at "large number of similar examples" might be "All examples with the same depth and label that are at a position > i"
* It might be worth keeping track of which `(depth, label)` pairs you've already done a bulk replacement on so you don't retry this multiple times if it failed, but the duplicate example detection is probably good enough.

I don't know how useful this will be for examples out in the wild, but I bet it will speed up our test suite!

**Comment by dchudz** [comment]
I'm not completely stuck, so my question probably won't stop me from looking at this more on my own. But I'd be very happy to get a little help.

Short version: 

- What's the philosophy around which `Example`s get labels? (Do we wish all of them did, or is leaving some with `None` part of the actual design?)

Long version:

I started looking at how `zero_draws` works for the case in `test_minimize_multiple_elements_in_silly_large_int_range`. Here's what the `Example`s list looks like:

```
[
Example(depth=0, label=None, start=0, end=701, discarded=False), 
 Example(depth=1, label=2038458451898798208, start=0, end=701, discarded=False), 
  Example(depth=2, label=2038458451898798208, start=0, end=701, discarded=False), 
   Example(depth=3, label=None, start=0, end=1, discarded=False), ***comes from start_example() in biased_coin()***
    Example(depth=4, label=None, start=0, end=1, discarded=False), 
   Example(depth=3, label=None, start=1, end=35, discarded=False), ***comes from start_example() in draw_bytes()***
    Example(depth=4, label=16793833830058185262, start=1, end=35, discarded=False), 
     Example(depth=5, label=None, start=1, end=2, discarded=False), 
     Example(depth=5, label=16949550145237255768, start=2, end=35, discarded=False), 
      Example(depth=6, label=10646957464545407928, start=2, end=35, discarded=False), 
       Example(depth=7, label=None, start=2, end=35, discarded=False), 
   Example(depth=3, label=None, start=35, end=36, discarded=False), 
    Example(depth=4, label=None, start=35, end=36, discarded=False), 
   Example(depth=3, label=None, start=36, end=70, discarded=False), 
    Example(depth=4, label=16793833830058185262, start=36, end=70, discarded=False), 
...
```

`zero_draws` can't reduce the 1st `depth=3` example because that's the biased coin that says to keep going, but it succeeds in reducing the 2nd `depth=3` example (and 4th, 6th, etc.).

I think implementing the suggestion exactly as written above ("All examples with the same depth and label that are at a position > i") wouldn't work, because:

- if `None`s are treated as having the same label, we'd attempt to zero out all of the `biased_coin()` draws and fail
- if `None`s are not treated as having the same label, we wouldn't attempt any bulk zero'ing (when we got to the `depth=4` example, it would already be zero'd, so we wouldn't do anything, and the bulk zero'ing wouldn't be triggered)

I can think of a few ways around this, but I'd love to have some more context for thinking about how to choose:

1. give everything labels (if all of these examples had meaningful labels, we'd have no problem)
2. attempt bulk-zero'ing even if the current example is already zero'd
3. attempt bulk-zero'ing if 
  - we just succeeded in zero'ing the current example and it has a label; or
  - ugly hack:
	  - the previous example succeeded in zero'ing
	  - we didn't attempt bulk-zero'ing for the previous example (e.g. because the previous example has no label)
	  - the current example does have a label
	  - the current example has the same start/end as the previous example
4. remove a layer of `Example` nesting (but I assume there's a good reason for `start_example()`/`stop_example()` -- just listing this for completeness)

(1) is my favorite of these, but I haven't really started thinking about how hard that is or why they don't currently get labeled.

**Comment by DRMacIver** [comment]
> give everything labels (if all of these examples had meaningful labels, we'd have no problem)

Giving every example a label would definitely be fine. Probably worth making that a separate pull request, but it should be fairly mechanical: Just remove the default from `start_example` and fix all of the `TypeError`s you get.

Sorry, for some reason  when writing up this suggestion I thought I'd done something sensible by default here when the label was None, but I clearly didn't.

**Comment by dchudz** [comment]
Thanks!

**Comment by Zac-HD** [comment]
Closed by #1843 🎉
